modules/javafx.controls/src/main/java/javafx/scene/chart/BarChart.java

```diff
@@ -209,8 +211,11 @@ public BarChart(@NamedArg("xAxis") Axis<X> xAxis, @NamedArg("yAxis") Axis<Y> yAx
         }
         // check if category is already present
         if (!categoryAxis.getCategories().contains(category)) {
+            // find category index in case data contains duplicate categories
+            int i = series.getData().size() != categoryAxis.getCategories().size() ? series.getItemIndex(item) :
```

Comment:
this is clever: little impact on performance in the case of non-duplicate categories (the most common case).

Reply:
However, I don't see the issue if I just add a duplicate data point after the chart is rendered (the code jumps to line 218 in the master branch).

This makes me think that perhaps the fix should include fixing the logic of adding category to the right hashtables at the right moment.  What do you think?

Reply:
I've added a popup menu on the BarChart page in the Monkey Tester to test the case ...
https://github.com/andy-goryachev-oracle/MonkeyTest

Reply:
You'll need to add an additional point after adding the one with duplicate categories to see the exception in the console. Using the "Add Point" button in the monkey tester after using the context menu entry should reproduce the issue.
Regarding category hashtables, do you have any specific issues in mind, or see any regressions due to this PR? Otherwise I'd suggest to create separate issues / failing tests and do follow-up PRs.

Reply:
The second added Point specifically needs to be in another, later category for this issue to occur - I created such a point in the monkey tester with the "Add Point" button to the right, not by repeatedly using the new context menu. Maybe try multiple times (depending on how the points random values are implemented there?)

Regarding categories: My understanding is that `series.getData()` is supposed to include all data, including duplicates. On the other hand, `categoryAxis.getCategories()` only contains distinct categories. So the new insertion index of the category of a (not necessarily duplicate) newly added point needs to be shifted only if there are any previous duplicates with lower indices. Thus, I think everything is accounted for correctly unless I'm missing something.